File: api/src/main/java/fr/huiitre/tools/modules/core/user/infrastructure/PostgresUserCredentialsRepository.java
package fr.huiitre.tools.modules.core.user.infrastructure;

import java.util.Optional;

import org.springframework.jdbc.core.JdbcTemplate;

import fr.huiitre.tools.modules.core.user.application.ports.UserCredentialsRepository;

public class PostgresUserCredentialsRepository implements UserCredentialsRepository {

    private final JdbcTemplate jdbcTemplate;

    public PostgresUserCredentialsRepository(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }

    @Override
    public void save(Long userId, String passwordHash) {
        final String sql = """
                    INSERT INTO tools_core.user_credentials (user_id, password_hash)
                    VALUES (?, ?)
                """;

        jdbcTemplate.update(sql, userId, passwordHash);
    }

    @Override
    public Optional<String> findPasswordHashByUserId(Long userId) {
        final String sql = """
                    SELECT password_hash
                    FROM tools_core.user_credentials
                    WHERE user_id = ?
                    LIMIT 1
                """;

        return jdbcTemplate
                .query(sql, rs -> rs.next() ? Optional.of(rs.getString("password_hash")) : Optional.empty(), userId);
    }

    @Override
    public void updatePassword(Long userId, String newPasswordHash) {
        final String sql = """
                    UPDATE tools_core.user_credentials
                    SET password_hash = ?
                    WHERE user_id = ?
                """;

        jdbcTemplate.update(sql, newPasswordHash, userId);
    }
}

File: api/Modules/Auth/Infrastructure/Persistence/PostgresUserCredentialsRepository.cs
using Dapper;
using Npgsql;
using Tools.Api.Modules.Auth.Application.Ports;
using Tools.Api.Modules.Common.Infrastructure;
using Tools.Api.Modules.Auth.Application.Ports.Password;

namespace Tools.Api.Modules.Auth.Infrastructure.Persistence;

// Adaptateur PostgreSQL/Dapper du port IUserCredentialsRepository.
public sealed class PostgresUserCredentialsRepository(PostgresSession session) : IUserCredentialsRepository
{
    public async Task<bool> ExistsAsync(long userId)
    {
        const string sql = "SELECT EXISTS (SELECT 1 FROM tools_core.user_credentials WHERE user_id = @UserId)";

        return await Connection().ExecuteScalarAsync<bool>(new CommandDefinition(
            sql, new { UserId = userId }, session.Transaction));
    }

    public async Task InsertAsync(long userId, string passwordHash)
    {
        const string sql = """
            INSERT INTO tools_core.user_credentials (user_id, password_hash)
            VALUES (@UserId, @PasswordHash)
            """;

        await Connection().ExecuteAsync(new CommandDefinition(
            sql,
            new { UserId = userId, PasswordHash = passwordHash },
            session.Transaction));
    }

    public async Task<int> UpdatePasswordAsync(long userId, string passwordHash)
    {
        const string sql = """
            UPDATE tools_core.user_credentials
            SET password_hash = @PasswordHash
            WHERE user_id = @UserId
            """;

        return await Connection().ExecuteAsync(new CommandDefinition(
            sql,
            new { UserId = userId, PasswordHash = passwordHash },
            session.Transaction));
    }

    private NpgsqlConnection Connection() => session.Connection
        ?? throw new InvalidOperationException("Aucune transaction PostgreSQL n'est ouverte.");
}
